test/template-tests/testcases/deploymentTemplate/IDs-Should-Be-Derived-From-ResourceIDs.test.ps1

```diff
@@ -1,16 +1,26 @@
-﻿param(
+﻿<#
+.Synopsis
+    Ensures that all IDs use the resourceID() function.
+.Description
+    Ensures that all IDs use the resourceID() function, or resolve to parameters or variables that use the ResourceID() function.
+.Example
+    Test-AzureRMTemplate -TemplatePath .\100-marketplace-sample\ -Test IDs-Should-Be-Derived-From-ResourceIDs
+.Example
+    .\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1 -TemplateObject (Get-Content ..\..\..\..\100-marketplace-sample\azureDeploy.json | ConvertFrom-Json) 
+#>
+param(
+# The template object (the contents of azureDeploy.json, converted from JSON)
 [Parameter(Mandatory=$true,Position=0)]
-$MainTemplateObject
-)
+$TemplateObject)
 
 # First, find all objects with an ID property in the MainTemplate.
-$ids = $MainTemplateObject  | Find-JsonContent -Key id -Value * -Like 
+$ids = $TemplateObject  | Find-JsonContent -Key id -Value * -Like
 
 foreach ($id in $ids) { # Then loop over each object with an ID
     $myId = "$($id.id)".Trim() # Grab the actual ID,
-    $expandedId = Expand-AzureRMTemplate -Expression $myId -InputObject $MainTemplateObject
-    if ($myId -notmatch '\[resourceId\(') { # check that it uses the ResourceID function
+    $expandedId = Expand-AzureRMTemplate -Expression $myId -InputObject $TemplateObject # then expand it.
+    if ($expandedId -notmatch '\[resourceId\(') { # Check that it uses the ResourceID function

```

Comment:
we also need to allow for 
[variables('
[parameters('

we may have to flip it to a blacklist instead, but for now, allow those three things at leat...

Reply:
Actually we've already got this.  That's what the Expand-AzureRMTemplate is bringing to the table.

If it is a "resolvable collection", e.g. a top-level property, Expand-AzureRMTemplate will expand out the expression.

Reply:
here's what I'm getting when I run on the 100-marketplace-sample

    Cannot convert value "IDs-Should-Be-Derived-From-ResourceIDs.test.ps1" to type "System.Management.Automation.CommandMetadata". Error: "At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:3 char:45                                  +     Ensures that all IDs use the resourceID() function.                                                                                                                +                                             ~                                                                                                                          An expression was expected after '('.                                                                                                                                                                                                                                                                                                             At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:5 char:45    +     Ensures that all IDs use the resourceID() function, or resolve to ...                                                                                              +                                             ~                                                                                                                          An expression was expected after '('.                                                                                                                                                                                                                                                                                                             At C:\Users\bmoore\source\repos\azure-quickstart-templates\test\template-tests\testcases\deploymentTemplate\IDs-Should-Be-Derived-From-ResourceIDs.test.ps1:5 char:119   + ... r resolve to parameters or variables that use the ResourceID() functi ...                                                                                          +                                                                  ~                                                                                                     An expression was expected after '('."